### PR TITLE
[PUBDEV-7203] Expose Individual Chunks via REST API

### PR DESCRIPTION
**Pull request by mn-mikke**
Lots of the code is extracted from [`ExternalFrameReaderBackend`](https://github.com/h2oai/h2o-3/blob/8b9882de0701538be731a0b4110dbec8fb67bf09/h2o-core/src/main/java/water/ExternalFrameReaderBackend.java) and [`ExternalFrameReaderClient`](https://github.com/h2oai/h2o-3/blob/914448614409b50b096893affa1b4a45905e0c12/h2o-core/src/main/java/water/ExternalFrameReaderClient.java) to minimalize changes. Once the whole initiative of REST transportation logic is done, all `ExternalFrame*`classes can be removed.

### CODE REVIEW

**Comment by mn-mikke** [comment]
Should i also add something like [DownloadDataHandler](https://github.com/h2oai/h2o-3/blob/43f8ab952a69a8bc9484bd0ffac909b6e3e820ca/h2o-core/src/main/java/water/api/DownloadDataHandler.java) and related things?

**Comment by michalkurka** [comment]
@mn-mikke why do you unpack the Chunk and then sent row-by-row? This means you are transferring much the data uncompressed, is it necessary to transfer row-by-row?

Wouldn't it be better to send compressed chunks and then reconstruct the rows (for iteration) on the client side (in SW)?

**Comment by mn-mikke** [comment]
@michalkurka Row-by-row data transfer would enable Sparkling Water to load only rows that are needed for processing. Chunk-by-chunk data transfer would enforce SW to load the whole partition into memory first and start with processing. To be honest with you, the original row-by-row data transfer loads the whole partition into memory anyway. I just wanted to say that sending data row by row could be the way how to optimize memory consumption in SW.

Is the chunk compression ratio significant?

File: h2o-core/src/main/java/water/ChunkAutoBufferWriter.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+public final class ChunkAutoBufferWriter implements Closeable  {
+    
+    private final AutoBuffer buffer;
+    
+    public ChunkAutoBufferWriter(OutputStream outputStream) {
+        this.buffer = new AutoBuffer(outputStream, false);
+    }
+
+    public void writeChunk(String frameName, int chunkId, byte[] expectedTypes, int[] selectedColumnIndices) {
+        Frame frame = DKV.getGet(frameName);
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/ChunkAutoBufferWriter.java]:
```suggestion
        final Frame frame = DKV.getGet(frameName);
```

Can be final

File: h2o-core/src/main/java/water/ChunkAutoBufferWriter.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        Chunk[] chunks = ChunkUtils.getChunks(frame, chunkId);
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/ChunkAutoBufferWriter.java]:
```suggestion
        final Chunk[] chunks = ChunkUtils.getChunks(frame, chunkId);
```

File: h2o-core/src/main/java/water/ChunkAutoBufferWriter.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        int numberOfRows = chunks[0]._len;
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/ChunkAutoBufferWriter.java]:
```suggestion
        final int numberOfRows = chunks[0]._len;
```

File: h2o-core/src/main/java/water/ChunkAutoBufferWriter.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        ExternalFrameUtils.sendInt(buffer, numberOfRows);
+
+        // buffered string to be reused for strings to avoid multiple allocation in the loop
+        BufferedString valStr = new BufferedString();
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/ChunkAutoBufferWriter.java]:
```suggestion
        final BufferedString valStr = new BufferedString();
```

File: h2o-core/src/main/java/water/ChunkAutoBufferWriter.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+public final class ChunkAutoBufferWriter implements Closeable  {
+    
+    private final AutoBuffer buffer;
+    
+    public ChunkAutoBufferWriter(OutputStream outputStream) {
+        this.buffer = new AutoBuffer(outputStream, false);
+    }
+
+    public void writeChunk(String frameName, int chunkId, byte[] expectedTypes, int[] selectedColumnIndices) {
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/ChunkAutoBufferWriter.java]: Do you really need to set the vectors types manually ? 

Introducing a function

```java
    private static byte[] vectorTypes(final Frame frame, final int[] vectors){
        final byte[] vectorTypes = new byte[vectors.length];

        for (int i = 0; i < vectors.length; i++) {
            vectorTypes[i] = frame.vec(i).get_type();
        }
        
        return vectorTypes;
    }
```

and then adding one line somewhere at the beginning of this method:

```java
final byte[] expectedTypes = vectorTypes(frame, selectedColumnIndices);
```

and then removing the expectedTypes would do the trick.

Comment (Pscheidl) [reply h2o-core/src/main/java/water/ChunkAutoBufferWriter.java]: Oh, I see, those are part of the client request. Do they serve for type checking, or for conversion as well ? If yes, what to do with incovertible types then ?

Do we check the types inside the frame are equal to the types requested ? Something like that.

```java
    private static boolean verifyVectorTypes(final Frame frame, final int[] vectors, final byte[] expectedTypes){

        for (int i = 0; i < vectors.length; i++) {
            if (expectedTypes[i] != frame.vec(vectors[i]).get_type()) return false;
        }
        
        return true;
    }
```

Comment (Pscheidl) [inline h2o-core/src/main/java/water/ChunkAutoBufferWriter.java]: Should we check the `expectedTypes` really match the types of vectors red from the Frame ? The type mapping has already been done by @jakubhava in `ExternalFrameUtils.vecTypesFromExpectedTypes`.

File: h2o-core/src/main/java/water/api/ChunkServlet.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  @Override
+  protected void doGet(HttpServletRequest request, HttpServletResponse response) {
+    String uri = ServletUtils.getDecodedUri(request);
+    try {
+      RequestParameters parameters = extractRequestParameters(request);
+      response.setContentType("application/octet-stream");
+      ServletUtils.setResponseStatus(response, HttpServletResponse.SC_OK);
+      OutputStream outputStream = response.getOutputStream();
+      ChunkAutoBufferWriter writer = new ChunkAutoBufferWriter(outputStream);
+      try {
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/api/ChunkServlet.java]: I suggest using try-with-resources block. 

```java
try(ChunkAutoBufferWriter writer = new ChunkAutoBufferWriter(response.getOutputStream())) {
```
1. The finally black may be deleted
2. The following lines above the try-catch block as well

```java
    OutputStream outputStream = response.getOutputStream();
      ChunkAutoBufferWriter writer = new ChunkAutoBufferWriter(outputStream);
```

File: h2o-core/src/main/java/water/api/ChunkServlet.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    buffer.get(selectedColumnIndices);
+            
+    return new RequestParameters(frameName, chunkId, expectedTypes, selectedColumnIndices); 
+  }
+  
+  @Override
+  protected void doGet(HttpServletRequest request, HttpServletResponse response) {
+    String uri = ServletUtils.getDecodedUri(request);
+    try {
+      RequestParameters parameters = extractRequestParameters(request);
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/api/ChunkServlet.java]:
```suggestion
      final RequestParameters parameters = extractRequestParameters(request);
```

File: h2o-core/src/main/java/water/api/ChunkServlet.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    IntBuffer buffer = ByteBuffer.wrap(selectedColumnsBytes).order(ByteOrder.BIG_ENDIAN).asIntBuffer();
+    int[] selectedColumnIndices = new int[buffer.remaining()];
+    buffer.get(selectedColumnIndices);
+            
+    return new RequestParameters(frameName, chunkId, expectedTypes, selectedColumnIndices); 
+  }
+  
+  @Override
+  protected void doGet(HttpServletRequest request, HttpServletResponse response) {
+    String uri = ServletUtils.getDecodedUri(request);
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/api/ChunkServlet.java]:
```suggestion
    final String uri = ServletUtils.getDecodedUri(request);
```

File: h2o-core/src/main/java/water/api/ChunkServlet.java [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    public RequestParameters(String frameName, int chunkId, byte[] expectedTypes, int[] selectedColumnIndices) {
+      this.frameName = frameName;
+      this.chunkId = chunkId;
+      this.expectedTypes = expectedTypes;
+      this.selectedColumnIndices = selectedColumnIndices;
+    }
+  }
+  
+  private String getParameterAsString(HttpServletRequest request, String parameterName) {
+    String result = request.getParameter(parameterName);
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/api/ChunkServlet.java]:
```suggestion
    final String result = request.getParameter(parameterName);
```

File: h2o-core/src/main/java/water/api/ChunkServlet.java [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    if (result == null) {
+      throw new RuntimeException(String.format("Cannot find value for parameter \'%s\'", parameterName));
+    }
+    return result;
+  }
+  
+  private RequestParameters extractRequestParameters(HttpServletRequest request) {
+    String frameName = getParameterAsString(request, "frame_name");
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/api/ChunkServlet.java]:
```suggestion
    final String frameName = getParameterAsString(request, "frame_name");
```

File: h2o-core/src/main/java/water/api/ChunkServlet.java [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    
+    String chunkIdString = getParameterAsString(request, "chunk_id");
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/api/ChunkServlet.java]:
```suggestion
    final String chunkIdString = getParameterAsString(request, "chunk_id");
```

File: h2o-core/src/main/java/water/api/ChunkServlet.java [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+    int chunkId = Integer.parseInt(chunkIdString);
+    
+    String expectedTypesString = getParameterAsString(request, "expected_types");
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/api/ChunkServlet.java]:
```suggestion
    final String expectedTypesString = getParameterAsString(request, "expected_types");
```

File: h2o-core/src/main/java/water/api/ChunkServlet.java [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+    byte[] expectedTypes = Base64.getDecoder().decode(expectedTypesString);
+
+    String selectedColumnsString = getParameterAsString(request, "selected_columns");
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/api/ChunkServlet.java]:
```suggestion
    final String selectedColumnsString = getParameterAsString(request, "selected_columns");
```

File: h2o-core/src/main/java/water/api/ChunkServlet.java [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+    byte[] selectedColumnsBytes = Base64.getDecoder().decode(selectedColumnsString);
+    IntBuffer buffer = ByteBuffer.wrap(selectedColumnsBytes).order(ByteOrder.BIG_ENDIAN).asIntBuffer();
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/api/ChunkServlet.java]:
```suggestion
    final IntBuffer buffer = ByteBuffer.wrap(selectedColumnsBytes).order(ByteOrder.BIG_ENDIAN).asIntBuffer();
```

File: h2o-core/src/main/java/water/api/ChunkServlet.java [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+            parameters.chunkId, 
+            parameters.expectedTypes,
+            parameters.selectedColumnIndices);
+      } finally {
+        writer.close();
+      }
+    } catch (Exception e) {
+      ServletUtils.sendErrorResponse(response, e, uri);
+    } finally {
+      ServletUtils.logRequest("GET", request, response);
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/api/ChunkServlet.java]:
```suggestion
      ServletUtils.logRequest(request.getMethod(), request, response);
```

File: h2o-core/src/main/java/water/ChunkAutoBufferWriter.java [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+
+import static water.ExternalFrameUtils.*;
+import static water.ExternalFrameUtils.EXPECTED_STRING;
+
+public final class ChunkAutoBufferWriter implements Closeable  {
+    
+    private final AutoBuffer buffer;
+    
+    public ChunkAutoBufferWriter(OutputStream outputStream) {
+        this.buffer = new AutoBuffer(outputStream, false);
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/ChunkAutoBufferWriter.java]:
```suggestion
        Objects.requireNonNull(outputStream);
        this.buffer = new AutoBuffer(outputStream, false);
```

File: h2o-core/src/main/java/water/ChunkAutoBufferWriter.java [hunk 17]
```diff
@@ -0,0 +1,89 @@
+package water;
+
+import water.fvec.Chunk;
+import water.fvec.ChunkUtils;
+import water.fvec.Frame;
+import water.parser.BufferedString;
+
+import java.io.Closeable;
+import java.io.IOException;
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/ChunkAutoBufferWriter.java]:
```suggestion
```

Unused import.

File: h2o-core/src/main/java/water/api/ChunkServlet.java [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+ * This servlet class handles GET requests for the path /3/Chunk
+ * It requires 4 get parameters
+ * - frame_name - a unique string identifier of H2O Frame
+ * - chunk_id - a unique identifier of the chunk within the H2O Frame
+ * - expected_type - byte array encoded in Base64 encoding. The types corresponds to the `selected_columns` parameter
+ * - selected_columns - selected columns indices encoded int Base64 encoding.
+ * The result is represented as a stream of binary data. Data are encoded to AutoBuffer row by row. 
+ * The data stream starts with the integer representing the number of rows.
+ */
+public class ChunkServlet extends HttpServlet {
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/api/ChunkServlet.java]:
```suggestion
public final class ChunkServlet extends HttpServlet {
```

File: h2o-core/src/main/java/water/api/ChunkServlet.java [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+    
+    public RequestParameters(String frameName, int chunkId, byte[] expectedTypes, int[] selectedColumnIndices) {
+      this.frameName = frameName;
+      this.chunkId = chunkId;
+      this.expectedTypes = expectedTypes;
+      this.selectedColumnIndices = selectedColumnIndices;
+    }
+  }
+  
+  private String getParameterAsString(HttpServletRequest request, String parameterName) {
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/api/ChunkServlet.java]:
```suggestion
  private String getParameterAsString(final HttpServletRequest request, final String parameterName) {
```

File: h2o-core/src/main/java/water/api/ChunkServlet.java [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+    String result = request.getParameter(parameterName);
+    if (result == null) {
+      throw new RuntimeException(String.format("Cannot find value for parameter \'%s\'", parameterName));
+    }
+    return result;
+  }
+  
+  private RequestParameters extractRequestParameters(HttpServletRequest request) {
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/api/ChunkServlet.java]:
```suggestion
  private RequestParameters extractRequestParameters(final HttpServletRequest request) {
```

File: h2o-core/src/main/java/water/ChunkAutoBufferReader.java [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+        this.buffer = new AutoBuffer(inputStream);
+        this.numRows = readInt();
+    }
+
+    public int getNumRows() {
+        return numRows;
+    }
+
+    public boolean readBoolean() {
+        boolean data = buffer.getZ();
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/ChunkAutoBufferReader.java]:
```suggestion
        final boolean data = buffer.getZ();
```

File: h2o-core/src/main/java/water/ChunkAutoBufferReader.java [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+        isLastNA = ExternalFrameUtils.isNA(buffer, data);
+        return data;
+    }
+
+    public byte readByte() {
+        byte data = buffer.get1();
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/ChunkAutoBufferReader.java]:
```suggestion
        final byte data = buffer.get1();
```

File: h2o-core/src/main/java/water/ChunkAutoBufferReader.java [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+        isLastNA = ExternalFrameUtils.isNA(buffer, data);
+        return data;
+    }
+
+    public char readChar() {
+        char data = buffer.get2();
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/ChunkAutoBufferReader.java]:
```suggestion
        final char data = buffer.get2();
```

File: h2o-core/src/main/java/water/ChunkAutoBufferReader.java [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+        isLastNA = ExternalFrameUtils.isNA(buffer, data);
+        return data;
+    }
+
+    public short readShort() {
+        short data = buffer.get2s();
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/ChunkAutoBufferReader.java]:
```suggestion
        final short data = buffer.get2s();
```

File: h2o-core/src/main/java/water/ChunkAutoBufferReader.java [hunk 25]
```diff
@@ -0,0 +1,93 @@
...
+        isLastNA = ExternalFrameUtils.isNA(buffer, data);
+        return data;
+    }
+
+    public int readInt() {
+        int data = buffer.getInt();
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/ChunkAutoBufferReader.java]:
```suggestion
        final int data = buffer.getInt();
```

File: h2o-core/src/main/java/water/ChunkAutoBufferReader.java [hunk 26]
```diff
@@ -0,0 +1,93 @@
...
+        isLastNA = ExternalFrameUtils.isNA(buffer, data);
+        return data;
+    }
+
+    public long readLong() {
+        long data = buffer.get8();
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/ChunkAutoBufferReader.java]:
```suggestion
        final long data = buffer.get8();
```

File: h2o-core/src/main/java/water/ChunkAutoBufferReader.java [hunk 27]
```diff
@@ -0,0 +1,93 @@
...
+        isLastNA = ExternalFrameUtils.isNA(buffer, data);
+        return data;
+    }
+
+    public float readFloat() {
+        float data = buffer.get4f();
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/ChunkAutoBufferReader.java]:
```suggestion
        final float data = buffer.get4f();
```

File: h2o-core/src/main/java/water/ChunkAutoBufferReader.java [hunk 28]
```diff
@@ -0,0 +1,93 @@
...
+        isLastNA = ExternalFrameUtils.isNA(data);
+        return data;
+    }
+
+    public double readDouble() {
+        double data = buffer.get8d();
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/ChunkAutoBufferReader.java]:
```suggestion
        final double data = buffer.get8d();
```

File: h2o-core/src/main/java/water/ChunkAutoBufferReader.java [hunk 29]
```diff
@@ -0,0 +1,93 @@
...
+        isLastNA = ExternalFrameUtils.isNA(data);
+        return data;
+    }
+
+    public String readString() {
+        String data = buffer.getStr();
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/ChunkAutoBufferReader.java]:
```suggestion
        final String data = buffer.getStr();
```

File: h2o-core/src/main/java/water/ChunkAutoBufferReader.java [hunk 30]
```diff
@@ -0,0 +1,93 @@
...
+        isLastNA = ExternalFrameUtils.isNA(buffer, data);
+        return data;
+    }
+
+    public Timestamp readTimestamp() {
+        Timestamp data = new Timestamp(buffer.get8());
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/ChunkAutoBufferReader.java]:
```suggestion
        final Timestamp data = new Timestamp(buffer.get8());
```

File: h2o-core/src/main/java/water/ChunkAutoBufferWriter.java (hunk 5 above)
Comment (mn-mikke) [reply h2o-core/src/main/java/water/ChunkAutoBufferWriter.java]: Great idea, thanks!

File: h2o-core/src/main/java/water/api/ChunkServlet.java (hunk 6 above)
Comment (mn-mikke) [reply h2o-core/src/main/java/water/api/ChunkServlet.java]: Thanks, will do.

**Comment by mn-mikke** [comment]
@Pscheidl thanks for your valuable comments, I've tried to address all of them. I also added validation of input parameters so may I ask you to take a look at the PR one more time?

**Comment by jakubhava** [comment]
@michalkurka we had sync up with @mn-mikke and Marek was explaining that you would prefer keeping the rest api only used by SW on Sparkling Water side.

This makes great sense, however then we would need to keep extending the jars which is the main thing we would like to get rid of. Ideally, the users should be able to use h2o drivers without any enhancements in this client-less scenarios.

Maybe we can do some compromise? We could create a separate repo, something like `sparkling-water-rest-api-extensions` where we would put these endpoints and H2O-3 assembly would depend on it, what do you think?

That way we would ensure H2O-3 does not need to maintain our specific endpoints, but it is included in the final builds of H2O

**Comment by jakubhava** [comment]
Just note to @michalkurka comment about unpacking the chunk: If we would send the chunk as they are, compressed, we could also remove our optimizations when sending the data (brings another complexity) and we could read data from chunks directly using methods available on chunks.

**Comment by michalkurka** [comment]
> @michalkurka we had sync up with @mn-mikke and Marek was explaining that you would prefer keeping the rest api only used by SW on Sparkling Water side.
> 
> This makes great sense, however then we would need to keep extending the jars which is the main thing we would like to get rid of. Ideally, the users should be able to use h2o drivers without any enhancements in this client-less scenarios.
> 
> Maybe we can do some compromise? We could create a separate repo, something like `sparkling-water-rest-api-extensions` where we would put these endpoints and H2O-3 assembly would depend on it, what do you think?
> 
> That way we would ensure H2O-3 does not need to maintain our specific endpoints, but it is included in the final builds of H2O

@jakubhava You could use `-libjars` to add this functionality dynamically and build the SW extension jar for H2O (wich all the SW specifics). For users launching manually, they would need to download SW distribution add add this jar (this is not common). For all other users - SW could add this automatically.

**Comment by michalkurka** [comment]
The other option is to move these SW extensions out of the h2o-core into a module of the h2o-3 repository.

**Comment by jakubhava** [comment]
@michalkurka I see, I forgot that actually Sparkling Water is the one who is responsible for creating the `hadoop jar` command.

I like this approach!. If @mn-mikke agrees, I would go with this solution as it will give us also more flexibility when it comes to developing changes related to external backend.

**Comment by mn-mikke** [comment]
The logic was moved to https://github.com/h2oai/sparkling-water/pull/1752. Closing the PR.

**Pull request closed without merging**